segmentation/utils/spatial_analysis.py

```diff
@@ -192,3 +197,72 @@ def calculate_cluster_spatial_enrichment(all_data, dist_mats, fovs=None,
+                cellLabelInImage (the cell label), and FlowSOM_ID (the cell phenotype id).
+            dist_matrices: A dictionary that contains a cells x cells matrix with the euclidian
+                distance between centers of corresponding cells for every fov
+            fovs: patient labels to include in analysis. If argument is none, default is all labels used.
+            distlim: cell proximity threshold. Default is 50.
+            fov_col: column with the cell fovs (Default is SampleID)
+            flowsom_col: column with the cell phenotype IDs (Default is FlowSOM_ID)
+            cell_label_col: column with the cell labels (Default is cellLabelInImage)
+            cell_type_col: column with the cell types (Default is cell_type)
+        Returns:
+            cell_neighbor_counts: matrix with phenotype counts per cell
+            cell_neighbor_freqs: matrix with phenotype frequencies of
+                counts per phenotype/total phenotypes for each cell
+            cell_count: current cell in analysis"""
+
+    # Setup input and parameters
+
+    if fovs is None:
+        fovs = sorted(list(set(all_data[fov_col])))
+
+    # Error Checking
+    if not np.isin(fovs, all_data[fov_col]).all():
+        raise ValueError("Points were not found in Expression Matrix")
+
+    # Get the phenotypes
+    pheno_titles = all_data[cell_type_col].drop_duplicates()
+
+    # Subset just the sampleID, cellLabelInImage, and FlowSOMID
+    all_data = all_data[[fov_col, cell_label_col, flowsom_col, cell_type_col]]
+    # Extract the columns with the cell phenotype codes
+    pheno_codes = all_data[flowsom_col].drop_duplicates()
+    # Get the total number of phenotypes
+    pheno_num = len(pheno_codes)
+
+    # initiate empty matrices for cell neighborhood data
+    cell_neighbor_counts = pd.DataFrame(np.zeros((all_data.shape[0], pheno_num + 2)))
+    cell_neighbor_freqs = cell_neighbor_counts.copy(deep=True)
+
+    # Replace the first and second columns of cell_neighbor_counts with the fovs and cell labels respectively
+    cell_neighbor_counts[[0, 1]] = all_data[[fov_col, cell_label_col]]
+    cell_neighbor_freqs[[0, 1]] = all_data[[fov_col, cell_label_col]]
+
+    # Rename the columns to match cell phenotypes
+    cols = [fov_col, cell_label_col] + list(pheno_titles)
+    cell_neighbor_counts.columns = cols
+    cell_neighbor_freqs.columns = cols
+
+    for i in range(len(fovs)):
+        # Subsetting expression matrix to only include patients with correct label
+        patient_ids = all_data.iloc[:, 0] == fovs[i]
+        fov_data = all_data[patient_ids]
+
+        # Subset the distance matrix dictionary to only include the distance matrix for the correct point
+        dist_matrix = dist_matrices[str(fovs[i])]
+
+        # Get cell_neighbor_counts and cell_neighbor_freqs for points
+        spatial_analysis_utils.compute_neighbor_counts(
```

Comment:
This might just be a pet peeve of mine, but I personally feel cell_neighbor_counts and cell_neighbor_freqs shouldn't be updated like this (as in, passing them into a function, and using the mutability properties of the parameters to update them in a separate function, in this case, compute_neighbor_counts). Let's see what Adam and Noah have to say, my suggestion is to initialize cell_neighbor_counts and cell_neighbor_freqs inside compute_neighbor_counts and having that function return those computed values.

Reply:
Updating mutables this way is generally considered good practice for a few reasons.  First, it guarantees that any changes to `cell_neighbor_counts` inside `compute_neighbor_counts` are never 'unnoticed' (e.g not overwritten on return, but used later).  Second, it should be a bit faster to directly update than to copy (shallowly or deeply) and then update with return values.  On the other hand, returning mutable objects should be done when the original input may be needed afterwards, and when you can guarantee that no changes to the original were made within the function.  (For a C/C++ analogy, think of it like passing by reference, versus passing by constant reference)

An examples of this in standard python is `random.shuffle(a)` which doesn't return a shuffled copy of `a` but directly updates it.

I recommend either leaving it as is, or (super optional) changing the order of arguments so that the mutables (`counts` and `freqs`) are the first two (could make it more clear that the other args are more like parameters, not true inputs).

Reply:
My main problem with this is that it requires you to go inside of compute_neighbor_counts to figure out that cell_neighbor_counts and cell_neighbor_freqs are secretly being updated.

Adam, thanks for the description, that's good to know. My understanding then is that because we know nothing else is going to be changes these mutables, it would be acceptable to return and update? 

Based on the way the function is currently doing the updating, it will require almost no changes to make this happen (see my comment inside of the helper function)

Reply:
when you put it like that, I think the best way to do this is to return the `counts` and `freqs` matrices and then update the `cell_neighbor_***` df's outside the function.  This way, those df's don't even need to be passed into the function, best practices aren't offended, and the code is mondo legible.